### DISCUSSION TITLE
Accessing jsPsych.timelineVariable() in the conditional_function()

### DISCUSSION THREAD

**Discussion by ClaraKuper** [jspsych/jsPsych#2022] [Q&A] [answered]
Hello everyone, 

I am currently programming an experiment where I want to repeat specific trials in a random order at the end of the timeline. 
I have trial and feedback sections (both handled in separate plugins), and I collect them in a common timeline with a loop function. This timeline repeats until all trials were successfully completed. If after one iteration of the timeline some trials were not completed, I don't want to repeat all trials, but only those that were not completed. I keep a list of IDs of all trials that have to be repeated. What I wanted to do is use the conditional_function() to check before each trial if the trial ID is in my list. The trial ID is saved as a timeline variable. 
The problem is the following:
It seems that in the conditional function, I don't have access to the timeline variables of the next trial - only to those from the last trial. (I tried calling jsPsych.timelineVariable('ID', true) - undefined - and jsPsych.currentTrial() -returns the last trial-).

How can I check a timeline variable of the upcoming trial in the conditional_function()? 

Thanks for your help!
Clara

DUMMY CODE:

trial = {
type: 'my-plugin',
ID: jsPsych.timelineVariable('ID'),
...,
on_finish: function(data){
if (not_completed){
repeat_IDs.push(sPsych.timelineVariable('ID'), true)
}

feedback = {
type: 'my-feedback',
...
}

trial_conditional = {
timeline: [trial],
conditional_function: CHECK HERE IF THE UPCOMING TRIAL ID IS PART OF REPEAT_IDs
}

feedback_conditional  = {
timeline: [feedback],
...
}

my_experiment = {
timeline: [trial_conditional, feedback_conditional],
timeline_variables: timeline_variables,
...
loop_function: function(){
return repeat_IDs.length>0
}

**Comment by becky-gilbert** [comment, marked as answer]
Hi @ClaraKuper, not sure if this is the best solution, but I think it works:
```js
  var timeline_variables = [
    {ID: 'trial 1'},
    {ID: 'trial 2'},
    {ID: 'trial 3'}
  ];

  var repeat_IDs = []; // holds list of IDs to repeat on the next round

  var first_set = true; // whether or not this is the first presentation of trials

  var trial = {
    type: 'html-keyboard-response',
    stimulus: jsPsych.timelineVariable('ID'),
    choices: jsPsych.ALL_KEYS,
    trial_duration: 2000,
    on_finish: function (data) {
      if (data.response == null) {
        if (!repeat_IDs.includes(jsPsych.timelineVariable('ID'))) {
          repeat_IDs.push(jsPsych.timelineVariable('ID'));
        }
      } else {
        repeat_IDs = repeat_IDs.filter(function(item) {
          return item !== jsPsych.timelineVariable('ID');
        });
      }
    }
  }

  var feedback = {
    type: 'html-keyboard-response',
    stimulus: 'feedback',
    trial_duration: 1000
  }

  var trial_conditional = {
    timeline: [trial, feedback],
    conditional_function: function() {
      if (first_set || repeat_IDs.includes(jsPsych.timelineVariable('ID'))) {
        return true;
      } else {
        return false;
      }
    }
  }

  var my_experiment = {
    timeline: [trial_conditional],
    timeline_variables: timeline_variables,
    loop_function: function () {
      first_set = false;
      return repeat_IDs.length > 0;
    }
  }
```
The thing that initially made this tricky was having separate conditional nodes for the trial and feedback, because the `repeat_IDs` array is used in both conditionals. This meant that you couldn't modify the `repeat_IDs` array at the end of the trial, because that would affect the feedback conditional. So I assumed that the trial and feedback will always occur or not occur for each stimulus, which is why I put the two trials in the same conditional node. The other important thing is to make sure to remove the stimulus ID from `repeat_IDs` if there is a response - I added this to the trial's `on_finish`. Without this bit, a stimulus ID will remain in the array even after it has been responded to, and the loop will continue forever.

Let us know if this works for you, and if you have any questions.

**Reply by ClaraKuper** [reply]
Hi @becky-gilbert  - thanks a lot for your reply! 

I tried to adapt my code to your suggestions, but the problem remained: I cannot access the value of timelineVariable('ID') in the conditional function. 

What is happening is the following:
Calling `jsPsych.timelineVariable('ID')` in the conditional function returns the function for a timeline variable check. repeat_IDs becomes an array of `function() { return timeline.timelineVariable(varname); }`.  The statement `repeat_IDs.includes(jsPsych.timelineVariable('ID'))` evaluates to false. This results in an infinite loop (the conditional function always evaluates to false, so the length of the repeat_IDs array is not changed and the loop function always evaluates to true) until the maximum call stack size is exceeded. I tried calling `jsPsych.timelineVariable('ID', true)` instead. That writes the trial ID into repeat_IDs during the on_finish function but during the conditional_function, this throws the following error: `TypeError: Cannot read property 'timelineVariable' of undefined`

I also implemented your script in an html document to see if the problem relates to my own code but it still persists.
I'll put the full html file below.

Any ideas how I could get the value of a timelineVariable at the moment of the conditional check? 

Thanks again,
Clara

```

<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Test repetition with jsPsych</title>
  <script src="jspsych-6.0.4/jspsych.js"></script>
  <script src="jspsych-6.0.4/plugins/jspsych-html-keyboard-response.js"></script>

</head>
<body>
<script>
let timeline_variables = [
    {ID: 'trial 1'},
    {ID: 'trial 2'},
    {ID: 'trial 3'}
  ];

let repeat_IDs = []; // holds list of IDs to repeat on the next round

let first_set = true; // whether or not this is the first presentation of trials

let trial = {
  type: 'html-keyboard-response',
  stimulus: jsPsych.timelineVariable('ID'),
    choices: jsPsych.ALL_KEYS,
    trial_duration: 2000,
    on_finish: function (data) {
      if (data.response == null) {
        if (!repeat_IDs.includes(jsPsych.timelineVariable('ID'))) {
          repeat_IDs.push(jsPsych.timelineVariable('ID'));
          console.log('added: ' + repeat_IDs);
        }
      } else {
        repeat_IDs = repeat_IDs.filter(function(item) {
          console.log('filtered:' + repeat_IDs);
          return item !== jsPsych.timelineVariable('ID');
        });
      }
    }
  }

let feedback = {
    type: 'html-keyboard-response',
    stimulus: 'feedback',
    trial_duration: 1000
  }

let trial_conditional = {
    timeline: [trial, feedback],
    conditional_function: function() {
      if (first_set || repeat_IDs.includes(jsPsych.timelineVariable('ID'))) { // ISSUE: jsPsych.timelineVariable('ID') returns a function (evaluates to false when compared to the list), calling jsPsych.timelineVariable('ID', true) throws an 'undefined' error 
        return true;
      } else {
        return false;
      }
    }
  }

let my_experiment = {
    timeline: [trial_conditional],
    timeline_variables: timeline_variables,
    loop_function: function () {
      first_set = false;
      return repeat_IDs.length > 0;
    }
  }

  jsPsych.init({
    timeline: [my_experiment], // use the timeline to run the experiment
    on_finish: function() {
      // print something to the console
      console.log('done');
    }
  });


</script>

</body>
</html>
```

**Reply by becky-gilbert** [reply]
Hi @ClaraKuper, ah, I see from your code above that you're using jsPsych v6.0.4. In earlier versions of jsPsych (prior to 6.3.0), you have to include the second `true` argument in the `jsPsych.timelineVariable` function whenever you want that function to return the variable's value (rather than returning _a function_ that returns the value, which is the default). Sorry, I should've mentioned this. 

So try adding the second `true` argument any time you use `jsPsych.timelineVariable` _inside of a function_, i.e. in the `on_finish` function for `trial`:
```js
let trial = {
  // other parameters...
  on_finish: function (data) {
    if (data.response == null) {
      if (!repeat_IDs.includes(jsPsych.timelineVariable('ID', true))) {  // here
        repeat_IDs.push(jsPsych.timelineVariable('ID', true));  // here
        console.log('added: ' + repeat_IDs);
      }
    } else {
      repeat_IDs = repeat_IDs.filter(function(item) {
        console.log('filtered:' + repeat_IDs);
        return item !== jsPsych.timelineVariable('ID', true);  // here
      });
    }
  }
}
```
And in the `conditional_function` for `trial_conditional`:
```js
let trial_conditional = {
    timeline: [trial, feedback],
    conditional_function: function() {
      if (first_set || repeat_IDs.includes(jsPsych.timelineVariable('ID', true))) { // here
        return true;
      } else {
        return false;
      }
    }
  }
```
You _don't_ want to use the second `true` argument when the `jsPsych.timelineVariable` function is used as the value of a trial parameter, like for your trial stimulus:
```js
let trial = {
  type: 'html-keyboard-response',
  stimulus: jsPsych.timelineVariable('ID'), // no true argument here
  // etc.
};
```
Does that work?

**Reply by ClaraKuper** [reply]
Hi @becky-gilbert, 

no, unfortunately with jsPsych 6.0.4 I still cannot call jsPsych.timelineVariable('ID', true) in the conditional function - I get an error that this property is undefined. 
But it was very easy to update to jsPSych 6.3.1 (latest version I found) - the compatibility is impeccable. I can now run your code above (with and without the "true" argument). Thank you!

**Reply by becky-gilbert** [reply]
Sorry that we couldn't figure out the problem with v6.0.4, but I'm glad to hear that upgrading to v6.3.1 fixed this for you! 

Keep in mind that there were quite a few changes made between these versions, for instance, the keyboard values switched from numbers (e.g. `choices: [78]`) to strings (e.g. `choices: ['f']`), preloading moved from `jsPsych.init` to a plugin, and some of parameter names and data properties have changed (e.g. button responses are now called `response` rather than `button_pressed` in the data). So if you do run into any other problems, it might be worth looking through the [release notes](https://github.com/jspsych/jsPsych/releases). And of course let us know if you have any questions or problems, and we'll do our best to help :smiley:
